### PR TITLE
Return the Position with the 'ad' available

### PR DESCRIPTION
**Pull request by alextremp**
This PR will:

* change the output contract of the OpenAds use cases returning the Position with the 'ad' attribute resolved, so there'll be no need to chain another promise in the client side.

Guys, this is a proposal to be merged, we can discuss about the need or not for doing this way :)

### CODE REVIEW

**Comment by victuxbb** [comment]
We need to re-think more this problem and go deeper in a better solution, there is no need to create a DTO response that is exactly the same as Domain entity, also wouldn't be a static method to create a DTO from Domain, we will have some service mapping.
IMHO we should face to face next points:
* Move proxy handler business to a kind of service or maybe inside AppnexusConsumerRepository
* Orchestration to obtain the adResponse from appnexus in use case and set it on domain entity

So finally all use cases respond with a domain entity with the property ad as a normal value.

**Comment by victuxbb** [comment]
Awesome job @alextremp

File: src/openads/application/service/AddPositionUseCase.js [hunk 1]
```diff
@@ -28,12 +32,29 @@ export default class AddPositionUseCase {
     return this._positionRepository.find({id})
       .then(this._filterPositionAlreadyExists)
       .then(() => this._positionFactory.create({id, name, source, placement, segmentation, sizes, native}))
-      .then(position => this._positionRepository.saveOrUpdate({position}))
+      // .then(this._setAdToPosition) // todo why this line fails ¿? the line below is working fine insted
+      .then(createdPosition => this._setAdToPosition(createdPosition))
```
Comment (victuxbb) [inline src/openads/application/service/AddPositionUseCase.js]: This is javascript faults, Promise is translated to es5 like in a global scope of window and doesnt recognize "this".
A possible solution is just bind this to the method like this:
```.then(this._setAdToPosition.bind(this))```

Comment (alextremp) [reply src/openads/application/service/AddPositionUseCase.js]: IMO the .bind(this) is a mindblower :) 
for readability I prefer, the arrow function option, is it OK for you?

Comment (victuxbb) [reply src/openads/application/service/AddPositionUseCase.js]: Yes no problem, but we agree on mix solutions? Some functions are pure and aren't using this, in that case we can use functions with one param like ```filterPositionAdIsAvailable```

Comment (alextremp) [reply src/openads/application/service/AddPositionUseCase.js]: 👍 Agree

File: src/openads/application/service/AddPositionUseCase.js [hunk 2]
```diff
@@ -28,12 +32,31 @@ export default class AddPositionUseCase {
     return this._positionRepository.find({id})
       .then(this._filterPositionAlreadyExists)
       .then(() => this._positionFactory.create({id, name, source, placement, segmentation, sizes, native}))
-      .then(position => this._positionRepository.saveOrUpdate({position}))
+      .then(createdPosition => this._setAdToPosition(createdPosition))
+      .then(positionWithAd => this._positionRepository.saveOrUpdate({position: positionWithAd}))
+      .then(savedPosition => this._filterPositionAdIsAvailable(savedPosition))
```
Comment (victuxbb) [inline src/openads/application/service/AddPositionUseCase.js]: we can make filterPositionAdIsAvailable as a function reference like filterPositionAlreadyExists

File: src/openads/application/service/DisplayPositionUseCase.js [hunk 3]
```diff
@@ -37,16 +37,14 @@ export default class DisplayPositionUseCase {
   }
 
   _filterPositionAdAvailable (position) {
-    return position.ad
-      .then(adResponse => adResponse.status)
+    return Promise.resolve(position.ad && position.ad.status)
```
Comment (victuxbb) [inline src/openads/application/service/DisplayPositionUseCase.js]: Niiiice!

File: src/openads/application/service/DisplayPositionUseCase.js [hunk 4]
```diff
@@ -37,16 +37,14 @@ export default class DisplayPositionUseCase {
...
       .then(status => AD_AVAILABLE === status)
-      .then(available => available ? position : Promise.reject(new PositionAdNotAvailableError({id: position.id})))
+      .then(available => available ? position : Promise.reject(new PositionAdNotAvailableError({position})))
   }
 
   _filterPositionAdNoNative (position) {
-    return position.ad
-      .then(adResponse => adResponse.data.adType)
+    return Promise.resolve(position.ad && position.ad.data && position.ad.data.adType)
```
Comment (victuxbb) [inline src/openads/application/service/DisplayPositionUseCase.js]: Fucking nice!

File: src/openads/infrastructure/repository/PullingAdRepository.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    this._positions.set(id, adResponse)
+  }
+
+  remove ({id}) {
+    this._positions.delete(id)
+  }
+
+  _waitForData ({id}) {
+    return Promise.resolve(this._positions.get(id))
+      .then(optionalPosition => optionalPosition || this._intervalPull(id))
```
Comment (victuxbb) [inline src/openads/infrastructure/repository/PullingAdRepository.js]: Awesome!

File: src/openads/application/service/AddPositionUseCase.js (hunk 2 above)
Comment (alextremp) [reply src/openads/application/service/AddPositionUseCase.js]: True! done :)

**Pull request merged**